worker/uniter/runner/context/context.go

```diff
@@ -313,7 +316,13 @@ func (ctx *HookContext) SetCacheValue(key, value string) error {
 		return err
 	}
 
+	curValue, exists := ctx.cacheValues[key]
+	if exists && curValue == value {
+		return nil // no-op
+	}
+
 	ctx.cacheValues[key] = value
+	ctx.cacheDirty = true
```

Comment:
It looks like there is a small gap in understanding if a real change has been made to the charm's saved state: a value is changed and returned to its original value before the data is flushed.   This approach appears to be best if we don't pull the data from state to verify a change before flush, or letting state determine that.

Reply:
If this scenario does happen (I believe it's rather unlikely but could be wrong) we will make an extra call to the server to persist the state and this will be a no-op server-side as we already check for state changes inside the update txn.

However, I don't anticipate this to be an issue; once we switch to a single transactional API call we will be making a call to the controller anyway when flushing.